dbms/src/DataStreams/ExpandBlockInputStream.cpp

```diff
@@ -0,0 +1,55 @@
+// Copyright 2023 PingCAP, Ltd.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include <DataStreams/ExpandBlockInputStream.h>
+#include <Interpreters/ExpressionActions.h>
+
+
+namespace DB
+{
+ExpandBlockInputStream::ExpandBlockInputStream(
+    const BlockInputStreamPtr & input,
+    const Expand2Ptr & expand2_,
+    const Block & header_,
+    const String & req_id)
+    : expand2(expand2_)
+    , header(header_)
+    , log(Logger::get(req_id))
+{
+    children.push_back(input);
+    i_th_project = 0;
```

Comment:
init `i_th_project` in init list

Reply:
Not exactly, when pipeline fetch a new block below Expand OP,  we should cache this new mother block and **rewind the current projection offset** for the convenience of later level projections (each time output a new projected block in function tryOutput(...))